Table writer 10: Support partition writes in TableWriter

gggrace14 opened the pull request:
When the table is partitioned, create multiple dwrf writers
by HiveDataSink, one for each distinct partition value.
Call PartitionIdGenerator to generate consecutive IDs
for distinct partition values, which are used for the index
of the dwrf writer vector. Each input could see new partition
values, so dwrf writers are created incrementally.

To assemble the input for different dwrf writers, we compute
the index into the original input for every partition value
and wrap the original input into a DictionaryVector.
The DictionaryVector is sent to the corresponding dwrf writer
as writer input. In the frequent case of the input belonging
to the same partition, no input wrap happens and we directly
send the input vector to the dwrf writer.

mbasmanova commented on velox/connectors/hive/HiveDataSink.h:
```diff
@@ -192,16 +192,96 @@ class HiveDataSink : public DataSink {
+  // partition values and the corresponding writers.
+  void moveWriters();
+
+  static RowVectorPtr makeWriterInput(
```
This method is a one-liner. It might be better to remove it and call "wrap" directly.

mbasmanova commented on velox/connectors/hive/HiveDataSink.h:
```diff
@@ -192,16 +192,96 @@ class HiveDataSink : public DataSink {
+  raw_vector<uint64_t> partitionIds_;
 };
 
+#define PARTITION_TYPE_DISPATCH(TEMPLATE_FUNC, typeKind, ...)             \
```
This macro is only used in the .cpp file. Can it be moved there?

It also looks very similar to VALUE_ID_TYPE_DISPATCH defined in VectorHasher.cpp. It would be nice to consolidate these two to avoid copy paste.

gggrace14 replied:
Moving the macro to .cpp file. To consolidate this macro w/ existing macros, I think I can directly call VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH, to avoid moving VALUE_ID_TYPE_DISPATCH from .cpp to .h.

mbasmanova commented on velox/connectors/hive/HiveConnector.h:
```diff
@@ -202,13 +202,22 @@ class HiveDataSource : public DataSource {
       "hive.immutable-partitions";
 
+  /// Maximum number of partitions per writer.
+  static constexpr const char* FOLLY_NONNULL kMaxPartitionsPerWriters =
```
Let's documents this configuration property in https://facebookincubator.github.io/velox/configs.html

mbasmanova commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -20,12 +20,81 @@
 
+namespace {
+std::vector<column_index_t> getPartitionChannels(
+    std::shared_ptr<const HiveInsertTableHandle> insertTableHandle) {
```
const &

mbasmanova commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -20,12 +20,81 @@
+  return channels;
+}
+
+RowTypePtr getChannelTypes(
```
Like I commented on the other PR, this looks should not be needed. We can pass 'rowType' to PartitionidGenerator directly.

gggrace14 replied:
Deprecated

mbasmanova commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -20,12 +20,81 @@
+
+uint64_t getMaxPartitionId(
+    const raw_vector<uint64_t>& partitionIds,
+    const SelectivityVector& activeRows) {
```
activeRows -> row; ditto other places

mbasmanova commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -20,12 +20,81 @@
+  return ROW(std::move(names), std::move(types));
+}
+
+uint64_t getMaxPartitionId(
```
I wonder if we can avoid this computation by fetching max partition ID from the ID generator.

gggrace14 replied:
We probably cannot.  I'm calling this getMaxPartitionId to get the max ID for current input, or the most recent input that is fed into the ID generator. On the other hand, I think we can only easily get the max ID from the ID generator or VectorHasher for ALL prior inputs. Max ID for current input might be smaller than max ID for all inputs globally.

mbasmanova replied:
@gggrace14 Is it important that max ID for the recent input and not global?

mbasmanova commented on velox/connectors/hive/HiveDataSink.h:
```diff
@@ -129,16 +191,97 @@ class HiveDataSink : public DataSink {
+      std::vector<vector_size_t>& rowCounts,
+      std::vector<BufferPtr>& rowIndices) const;
+
+  static std::string getPartitionString(
```
This method can be a function in anonymous namespace in the .cpp file.

gggrace14 replied:
Moved

mbasmanova reviewed (commented):
@gggrace14 I'm still reading this PR, but here are some initial comments.

mbasmanova commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -34,33 +103,138 @@ HiveDataSink::HiveDataSink(
-  if (writers_.empty()) {
-    writers_.emplace_back(createWriter());
+  input_ = std::dynamic_pointer_cast<RowVector>(input);
+  input_->loadedVector();
```
input is never a lazy vector. Hence, calling loadedVector is redundant. However, child vectors might be lazy. Hence, you need to call loadedVector on them.

Here is an example from HashProbe.cpp:

```
  // Ensure vector are lazy loaded before spilling.
  for (int32_t i = 0; i < input_->childrenSize(); ++i) {
    input_->childAt(i)->loadedVector();
  }
```

gggrace14 replied:
Got you, good to know. I change to call loadedVector() on individual child vectors of input_, and it works!

mbasmanova commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -70,22 +244,151 @@ std::unique_ptr<velox::dwrf::Writer> HiveDataSink::createWriter() {
+    writerParameters_.resize(maxNewIndex + 1);
+    partitionsVector_->resize(maxNewIndex + 1, false);
+  }
+  // Rehash monotonically makes the IDs of previous partition values larger. So
```
This assumption tightly couples this code with VectorHasher's implementation. It would be nice to avoid that. Alternatively, we'd need to add unit tests that assert this behavior, so that if it ever changes, we'll get a clear signal.

gggrace14 replied:
I chose to add unit tests to assert this behavior. W/o this assumption an alternative approach would be to have a second vector when moving the writers. Decided to trade away robustness for performance here.

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -34,33 +103,138 @@ HiveDataSink::HiveDataSink(
+  toCommit.reserve(writerParameters_.size());
+
+  for (auto i = 0; i < writerParameters_.size(); i++) {
+    if (writerParameters_[i] != nullptr) {
```
Some indices of writerParameters_ can be nullptr, as the IDs from the PartitionIdGenerator or VectorHasher are not consecutive.

mbasmanova replied:
Thank you for clarifying.

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -34,33 +103,138 @@ HiveDataSink::HiveDataSink(
+  input_->loadedVector();
+
+  auto numRows = input->size();
+  auto numPartitions = partitionChannels_.size();
```
You're right. Remove this local variable as it's not used many times. It's from legacy code change.

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -34,33 +103,138 @@ HiveDataSink::HiveDataSink(
+              writeProtocol_->getWriterParameters(
+                  insertTableHandle_,
+                  connectorQueryCtx_,
+                  std::make_shared<HiveConnectorWriteInfo>(std::nullopt)));
```
Good to know.

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -34,33 +103,138 @@ HiveDataSink::HiveDataSink(
+  // Writing to partitioned table
+  activeRows_.resize(numRows);
+  activeRows_.setAll();
+  bool stable = partitionIdGenerator_.run(input_, activeRows_, partitionIds_);
```
Yes, SelectivityVector is set to a subset of rows when we apply partitionIdGenerator_ to the partitionsVector_, in moveWriters(), to recalculate the IDs for previous partition values. Not all rows of partitionsVector_ are set as the IDs generated from partitionIdGenerator_ are not consecutive.

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -34,33 +103,138 @@ HiveDataSink::HiveDataSink(
+    moveWriters();
+  }
+
+  uint64_t maxPartitionId = getMaxPartitionId(partitionIds_, activeRows_);
```
Removing activeRows_ from the API.

Regarding asking partitionIdGenerator_ for the max, commented the same as above
We probably cannot. I'm calling this getMaxPartitionId to get the max ID for current input, or the most recent input that is fed into the ID generator. On the other hand, I think we can only easily get the max ID from the ID generator or VectorHasher for ALL prior inputs. Max ID for current input might be smaller than max ID for all inputs globally.

gggrace14 commented on velox/connectors/hive/HiveDataSink.h:
```diff
@@ -129,16 +191,97 @@ class HiveDataSink : public DataSink {
+  // partition ID based on the ID labeling of rowPartitionIds. Both rowCounts
+  // and rowIndices are indexed by partition ID. rowIndices[i] is a vector of
+  // all rows labeled with partition ID i.
+  static void computePartitionRowCountsAndIndices(
```
Moving it to anonymous namespace

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -70,22 +244,151 @@ std::unique_ptr<velox::dwrf::Writer> HiveDataSink::createWriter() {
+// static
+void HiveDataSink::computePartitionRowCountsAndIndices(
+    const raw_vector<uint64_t>& rowPartitionIds,
+    const SelectivityVector& activeRows,
```
Deprecating the SelectivityVector.

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -34,33 +103,138 @@ HiveDataSink::HiveDataSink(
+
+  uint64_t maxPartitionId = getMaxPartitionId(partitionIds_, activeRows_);
+
+  std::vector<BufferPtr> partitionRowIndices;
```
Revised

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -70,22 +244,151 @@ std::unique_ptr<velox::dwrf::Writer> HiveDataSink::createWriter() {
+  rowCounts.resize(numPartitionIds);
+  std::fill(rowCounts.begin(), rowCounts.end(), 0);
+  rowIndices.resize(numPartitionIds);
+  std::fill(rowIndices.begin(), rowIndices.end(), nullptr);
```
Revised to reuse the vectors & buffers across batches. So keeping these reset calls.

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -34,33 +103,138 @@ HiveDataSink::HiveDataSink(
+    writerParameters_.resize(maxPartitionId + 1);
+  }
+
+  partitionsVector_->copyRanges(
```
I guess the copy-single-row API is the copy() API as below
https://github.com/facebookincubator/velox/blob/eaf59f1f974cb036d7a8f1593a06bc0e9255fb0e/velox/vector/BaseVector.h#L401-L422

I also read from the above link that copying multiple rows at once is more efficient than copying single rows multiple times. So chose to prepare the Range that covers all the new partitions and call the copyRanges() API. Let me know if my understanding is off.

mbasmanova replied:
It is true that copying ranges is more efficient than copying one row at a time, but only if you are copying large number of rows. Here, we will be copying a handful of rows each time, hence, code simplicity is more important.

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -70,22 +244,151 @@ std::unique_ptr<velox::dwrf::Writer> HiveDataSink::createWriter() {
+
+  for (vector_size_t id = 0; id < rowCounts.size(); id++) {
+    if (rowCounts[id] > 0 && !writers_[id]) {
+      copyRanges.push_back(
```
Copying comment from above

I also read from the above link that copying multiple rows at once is more efficient than copying single rows multiple times. So chose to prepare the Range that covers all the new partitions and call the copyRanges() API. Let me know if my understanding is off or it is not necessary.

gggrace14 replied:
Took this suggestion and revised.

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -70,22 +244,151 @@ std::unique_ptr<velox::dwrf::Writer> HiveDataSink::createWriter() {
 }
 
+void HiveDataSink::moveWriters() {
+  raw_vector<uint64_t> newIndices;
```
Got it. Revised

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -34,33 +103,138 @@ HiveDataSink::HiveDataSink(
+    if (partitionRowCounts[partitionId] == 0) {
+      continue;
+    }
+    if (!writers_[partitionId]) {
```
Created a helper func and removed the code repetition.

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -70,22 +244,151 @@ std::unique_ptr<velox::dwrf::Writer> HiveDataSink::createWriter() {
+  // All valid values of partitionsVector_ are from previous addInput() that
+  // have been fed into partitionIdGenerator at least once, and thus assumed not
+  // to cause partitionIdGenerator to rehash.
+  partitionIdGenerator_.stableRun(partitionsVector_, validRows, newIndices);
```
Nope, validRows has a subset of rows selected. Not all rows of partitionsVector_ are selected since the index of partitionsVector_ generated from VectorHasher are not consecutive.

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -34,33 +103,138 @@ HiveDataSink::HiveDataSink(
+  }
+
+  partitionsVector_->copyRanges(
+      makeInputPartitionsVector(input_).get(),
```
Took the suggestion to iterate over only the partition columns and copy by a single row. I put the logic into the createWriterForPartition() func.

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -34,33 +103,138 @@ HiveDataSink::HiveDataSink(
+      std::string partitionString =
+          getPartitionString(partitionsVector_, partitionId);
+
+      auto writerParameters =
```
Created a helper func and removed the code repetition.

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -70,22 +244,151 @@ std::unique_ptr<velox::dwrf::Writer> HiveDataSink::createWriter() {
+void HiveDataSink::moveWriters() {
+  raw_vector<uint64_t> newIndices;
+
+  SelectivityVector validRows = SelectivityVector::empty(writers_.size());
```
At this place we need to get a SelectivityVector of the valid rows of the partitionsVector_ and pass it to partitionIdGenerator_, as not all element of partitionsVector_ are set. Originally I use the writers_ to infer the SelectivityVector, using the writer_ as the source of truth. As it sounds confusing, now I change to have a private field, SelectivityVector validPartitionsVectorRows_, to track the valid rows of partitionsVector_ and update it at the same time we update partitionsVector_. Logic is in moveWriters() and createWriterForPartition().

mbasmanova commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -17,15 +17,136 @@
+    const std::string& name) {
+  using T = typename TypeTraits<Kind>::NativeType;
+  auto* simpleVector = partitionVector->as<SimpleVector<T>>();
+  VELOX_CHECK_NOT_NULL(simpleVector, "");
```
Error message is empty string. Is this intentional?

gggrace14 replied:
Nope :(

mbasmanova commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -17,15 +17,136 @@
+      makeColumnString(partitionVector->as<SimpleVector<T>>()->valueAt(row)));
+};
+
+std::string getPartitionString(
```
nit: Would it make sense to add this function to a header file and add a unit test?

gggrace14 replied:
Will do.

mbasmanova commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -34,33 +155,100 @@ HiveDataSink::HiveDataSink(
     const {
-  return std::make_shared<HiveConnectorCommitInfo>(writerParameters_);
+  std::vector<std::shared_ptr<const HiveWriterParameters>> toCommit;
+  toCommit.reserve(writerParameters_.size());
```
Should we use numPartitions_ here?

gggrace14 replied:
You're right, numPartitions_ is better.

mbasmanova commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -34,33 +155,100 @@ HiveDataSink::HiveDataSink(
+      VELOX_CHECK_LE(
+          numPartitions_,
+          HiveConfig::maxPartitionsPerWriters(connectorQueryCtx_->config()),
+          "Exceeded limit of {} open writers for partitions",
```
The error message don't need to include the limit. VELOX_CHECK_LE will include numPartition_ and limit automatically.

```
VELOX_CHECK_LE(
          numPartitions_,
          HiveConfig::maxPartitionsPerWriters(connectorQueryCtx_->config()),
          "Exceeded limit of open writers for partitions");
```

Do we have a test that triggers this error?

gggrace14 replied:
Ah, I see. Kind of saw this before.

Since we're going to move the limit check to the ID generator class, I will add a unit test there for the limit check.

mbasmanova replied:
Thanks. Let's also make sure to use VELOX_USER_CHECK_LE to generate user exception, not system exception.

gggrace14 replied:
Makes sense to me

mbasmanova commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -70,22 +258,63 @@ std::unique_ptr<velox::dwrf::Writer> HiveDataSink::createWriter() {
+  // Rehash monotonically makes the IDs of previous partition values larger. So
+  // moving the existing writers to the high indices starting with the high
+  // end is safe and will not overwrite existing ones.
+  for (vector_size_t i = writers_.size() - 1; i >= 0; i--) {
```
nit: Any particular reason not to loop over idMap like you do in the ID generator?

gggrace14 replied:
You're right. That'll cost less.

mbasmanova reviewed (commented):
@gggrace14 Looks good. Some follow-up questions.

gggrace14 commented on velox/connectors/hive/HiveWriteProtocol.h:
```diff
@@ -70,6 +78,10 @@ class HiveWriterParameters : public WriterParameters {
     }
   }
 
+  const std::optional<std::string>& partitionName() const {
```
HiveWriterParameters::partitionName() can be retrieved in HiveWriteProtocol::commit() and be returned as output of the TableWriter and be passed to the next operator. Here is how we use it in the presto_cpp WriteProtocol, https://github.com/prestodb/presto/pull/18810/files.

gggrace14 commented on velox/connectors/hive/HiveConfig.h:
```diff
@@ -0,0 +1,95 @@
+namespace facebook::velox::connector::hive {
+
+/// Hive connector configs
+class HiveConfig {
```
Added the doc. I add a section for Hive configuration properties. Existing config properties are query configuration properties.

gggrace14 commented on velox/connectors/hive/HiveConfig.h:
```diff
@@ -0,0 +1,95 @@
+  }
+
+ private:
+  static InsertExistingPartitionsBehavior
```
I see.

gggrace14 commented on velox/connectors/hive/HiveConfig.h:
```diff
@@ -0,0 +1,95 @@
+
+  static InsertExistingPartitionsBehavior insertExistingPartitionsBehavior(
+      const Config* FOLLY_NONNULL baseConfig) {
+    auto strBehavior =
```
Revised

gggrace14 commented:
Not ready for review yet. Will do more revising.

gggrace14 commented:
@mbasmanova Hi Masha, I revised this PR based on our discussion that utilizes the ID generator based on only a single VectorHasher. Also addressed the prior comments that are applied to the revision. Would you be able to help take a look again?

mbasmanova commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -17,56 +17,121 @@
+      createWriterForPartition(partitionId);
+    }
+
+    RowVectorPtr writerInput = exec::wrap(
```
Writing a single partition is a common use case. Consider optimizing for that (can be a follow-up PR).

mbasmanova commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -117,15 +208,52 @@ void HiveDataSink::createWriter(
+  partitionSizes_.resize(numPartitionIds);
+  std::fill(partitionSizes_.begin(), partitionSizes_.end(), 0);
+  partitionRows_.resize(numPartitionIds);
+  std::fill(partitionRows_.begin(), partitionRows_.end(), nullptr);
```
This code prevents the reuse of the buffers. Please, revise.

gggrace14 replied:
Revised. The individual Buffers of partitionRows_ need to resize to have more capacity if needed. What functions shall I call to increase capacity?

I'm using
```c++
partitionRows_[id] = allocateIndices(numRows, pool);
```
now to reallocate the memory. Do I need to call some function to release the memory of the old Buffer?

mbasmanova replied:
Check out Buffer::reallocate method.

mbasmanova commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -117,15 +208,52 @@ void HiveDataSink::createWriter(
+    uint64_t id = partitionIds_[row];
+
+    if (partitionRows_[id] == nullptr) {
+      partitionRows_[id] = allocateIndices(numRows, pool);
```
It would be nice to skip populating this buffer when writing to a single partition. PartitionIdGenerator should be able to tell us whether there is only one partition.

gggrace14 replied:
Yeah, revised to take a bool for single partition and skip populating the buffer if the bool is true.

gggrace14 replied:
Hi @mbasmanova Masha, here I need to figure out a way to tell if it is writing to a single partition.

You suggest we should be able to get this info from PartitionIdGenerator, and I get it. The way I'm trying here is to keep track of the recentMaxPartitionId_ and recentMinPartitionId_ inside PartitionIdGenerator, and compare them to see if there's only a single partition id. Inside PartitionIdGenerator we traverse the rows of input anyway, so keeping track of max & min is not adding more traversal overhead. Also we need to expose recentMaxPartitionId.

https://github.com/facebookincubator/velox/blob/d3accd7b4066ab7b6eafef1e2e25e553ef8805c5/velox/connectors/hive/PartitionIdGenerator.cpp#L83-L84

Also in your previous PR, you have this comment,
https://github.com/facebookincubator/velox/blob/4653ec4329b4366f392c5519e1ea3728b337e1e3/velox/connectors/hive/PartitionIdGenerator.cpp#L61-L63

I think here we need to tell if the current input writes to a single partition, while looking at every VectorHasher's # of UniqueValues gives us if all inputs in history write to a single partition, which is more strict.

Let me know what you think.

mbasmanova commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -103,12 +193,13 @@ std::vector<std::string> HiveDataSink::finish() const {
 
-void HiveDataSink::createWriter(
-    const std::optional<std::string>& partitionName) {
+void HiveDataSink::createWriter(vector_size_t partitionId) {
```
Any particular reason not to pass partition name along with partition Id?

I noticed that you add input_ as a member variable, but this should not be necessary. The caller can compute partition name and pass it in.

gggrace14 replied:
input_ member variable was removed. And moved partitionName definition out of createWriter()

mbasmanova commented on velox/connectors/hive/HiveDataSink.h:
```diff
@@ -205,10 +214,21 @@ class HiveDataSink : public DataSink {
+  // Below are structures updated when generating partition IDs for the current
+  // input. input_ and partitionIds_ are indexed by the row of input_.
+  // partitionRows_ and partitionSizes_ are indexed by partitionId.
+  RowVectorPtr input_;
```
This member variable should not be needed.

gggrace14 replied:
Removed

mbasmanova commented on velox/connectors/hive/HivePartitionUtil.h:
```diff
@@ -0,0 +1,96 @@
+          partitionVector->as<SimpleVector<T>>()->valueAt(row)));
+};
+
+inline std::string makePartitionName(
```
Why 'inilne'? Consider moving the implementation of this function to a .cpp file.

It would be nice to extract this function and test into a separate PR.

gggrace14 replied:
Moved to a separate .cpp file

mbasmanova commented on velox/connectors/hive/HivePartitionUtil.h:
```diff
@@ -0,0 +1,96 @@
+}
+
+template <TypeKind Kind>
+inline std::string makePartitionKeyValueString(
```
Move this code to .cpp file.

mbasmanova commented on velox/connectors/hive/HivePartitionUtil.h:
```diff
@@ -0,0 +1,96 @@
+    const std::string& name) {
+  using T = typename TypeTraits<Kind>::NativeType;
+  auto* simpleVector = partitionVector->as<SimpleVector<T>>();
+  VELOX_CHECK_NOT_NULL(
```
This check is not needed. SimpleVector is the base class for all encodings. Hence, this check cannot fail.

gggrace14 replied:
Got it. Removed

mbasmanova commented on velox/connectors/hive/tests/HivePartitionUtilTest.cpp:
```diff
@@ -0,0 +1,79 @@
+    std::vector<column_index_t> partitionChannels{0, 1, 2, 3, 4, 5, 6};
+
+    EXPECT_EQ(
+        makePartitionName(input, partitionChannels, 0),
```
Perhaps, add a test case with a single partition key. That's the most common case.

gggrace14 replied:
Revised the test to include the case of single partition.

mbasmanova commented on velox/connectors/hive/tests/HivePartitionUtilTest.cpp:
```diff
@@ -0,0 +1,79 @@
+         "flat_tinyint_col",
+         "flat_smallint_col",
+         "flat_int_col",
+         "flat_bigint_col",
```
Perhaps, add DATE column as well.

mbasmanova commented on velox/connectors/hive/tests/HivePartitionUtilTest.cpp:
```diff
@@ -0,0 +1,79 @@
+  template <typename T>
+  VectorPtr makeDictionary(const std::vector<T>& data) {
+    auto base = makeFlatVector(data);
+    auto indices = makeIndices(data.size(), [](auto row) { return row; });
```
This dictionary is not interesting. Consider testing constant vectors as well as dictionary vectors where a few values repeat many times.

gggrace14 replied:
Revised for dictionary vector. Testing on constant vector is included as the last column in the test case.

mbasmanova commented on velox/exec/tests/TableWriteTest.cpp:
```diff
@@ -40,12 +40,39 @@ class TableWriteTest : public HiveConnectorTestBase {
     return splits;
   }
 
+  uint32_t fileCount(const std::shared_ptr<TempDirectoryPath>& directoryPath) {
```
nit: fileCount -> countFilesRecursively?

gggrace14 replied:
Revised

mbasmanova commented on velox/exec/tests/TableWriteTest.cpp:
```diff
@@ -264,6 +294,115 @@ TEST_F(TableWriteTest, TestASecondCommitStrategy) {
+      makeHiveConnectorSplits(outputDirectory),
+      "SELECT * FROM tmp");
+
+  EXPECT_EQ(
```
Can we make this verification stronger and check the directory structure expecting one sub-directory per partition (with the right name) and one file in each sub-directory (with one row in each file)?

gggrace14 replied:
Previously I tried similar but couldn't achieve it with what we can get from the std filesystem library, https://en.cppreference.com/w/cpp/filesystem.

Ex., the difficulty is, for partitions a=1/b=2 and a=1/b=3, the fs::recursive_directory_iterator() gives a list of
a=1
a=1/b=2
a=1/b=2/file2
a=1/b=3
a=1/b=3/file3

I can't tell the directory a=1 apart from the rest.

mbasmanova commented on velox/exec/tests/TableWriteTest.cpp:
```diff
@@ -264,6 +294,115 @@ TEST_F(TableWriteTest, TestASecondCommitStrategy) {
+      "SELECT * FROM tmp");
+
+  EXPECT_EQ(
+      fileCount(outputDirectory), numPartitions * task->numOutputDrivers());
```
What is the value of task->numOutputDrivers() here?

gggrace14 replied:
`task->numOutputDrivers()` is just used to capture the # of drivers for the TableWriter. As you might already know, it's set here and now is 1.
https://github.com/facebookincubator/velox/blob/3bb9726871b3901f9fb37d61480f082a7c4892ee/velox/exec/LocalPlanner.cpp#L215-L221

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -45,6 +56,39 @@ std::string makePartitionDirectory(
+  return boost::lexical_cast<std::string>(boost::uuids::random_generator()());
+}
+
+std::string makePartitionUpdate(
```
Removed

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -55,17 +99,63 @@ HiveDataSink::HiveDataSink(
+      createWriter(partitionId);
+    }
+
+    RowVectorPtr writerInput = exec::wrap(
```
Revised.

mbasmanova commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -118,14 +190,55 @@ void HiveDataSink::createWriter(
+      partitionRows_[id] = allocateIndices(numRows, pool);
+    }
+
+    auto* rawIndex = partitionRows_[id]->asMutable<vector_size_t>();
```
Consider storing raw pointers to these buffers to avoid casting for each row.

gggrace14 replied:
Borrow it from your example

mbasmanova reviewed (commented):
@gggrace14 Some follow-up comments.

gggrace14 commented on velox/connectors/hive/HiveConfig.h:
```diff
@@ -31,8 +31,14 @@ class HiveConfig {
   static InsertExistingPartitionsBehavior insertExistingPartitionsBehavior(
       const Config* config);
+
+  static uint32_t maxPartitionsPerWriters(const Config* baseConfig);
```
Sorry, revised to `config`. Guess it might be from back and forth copying.

I just found out configs.rst already has the description for max_partitions_per_writers. I think it's because I forgot to remove it from the previous PR :(.

gggrace14 commented on velox/connectors/hive/PartitionIdGenerator.h:
```diff
@@ -46,6 +46,11 @@ class PartitionIdGenerator {
   }
 
+  /// Return the minimum partition ID generated for the most recent input.
+  bool hasSinglePartitionId() const {
```
This func is gone

gggrace14 commented on velox/exec/tests/TableWriteTest.cpp:
```diff
@@ -40,12 +40,40 @@ class TableWriteTest : public HiveConnectorTestBase {
     return splits;
   }
 
+  uint32_t countFilesRecursively(
```
Unit tests still to be revised

mbasmanova replied:
@gggrace14 Thank you, Ge. Let's makes sure to verify the generated directory names, file content, single and multiple partition keys, single and multiple partitions, a limit on total number of partitions.

gggrace14 replied:
Test cases are updated to cover generated directory names, file content, single and multiple partition keys, single and multiple partitions. For the limit on the total number of partitions, we have a test case for PartitionIdGenerator, which is where the limit is enforced. Let me know if we also want one test case for TableWriteTest. https://github.com/facebookincubator/velox/blob/2e6f23fb4cf17e6dce61d44b7b3a464c0076c5f0/velox/connectors/hive/tests/PartitionIdGeneratorTest.cpp#L142-L156

mbasmanova replied:
@gggrace14 It would be nice to have an e2e limit test as well to ensure TableWriter properly fetches the limit from the config and passes it to PartitionIdGenerator.

gggrace14 commented:
Hi @mbasmanova , the PR is updated and ready for review. I borrowed the example piece of code you put together for me, w/ minor adjustment, and found that ensurePartitionWriters() helps largely simplify the code logic flow.

mbasmanova commented:
@gggrace14 Ge, there is a CircleCI failure which seems to be due to changes in this PR. Would you take a look?

gggrace14 commented:
> @gggrace14 Ge, there is a CircleCI failure which seems to be due to changes in this PR. Would you take a look?

Hi @mbasmanova Masha, I fixed the CircleCI failure. Also added a TableWriteTest case to test the limit on partition # end to end :)

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -118,14 +190,55 @@ void HiveDataSink::createWriter(
+
+  partitionSizes_.resize(numPartitionIds);
+  std::fill(partitionSizes_.begin(), partitionSizes_.end(), 0);
+  partitionRows_.resize(numPartitionIds, nullptr);
```
Learned

mbasmanova commented on velox/exec/tests/TableWriteTest.cpp:
```diff
@@ -228,40 +291,258 @@ TEST_F(TableWriteTest, constantVectors) {
+        fmt::format(
+            "SELECT * FROM tmp WHERE {}",
+            partitionNameToPredicate(*iterPartitionName)));
+    // Files written to each partition directory is per driver.
```
What is 'task->numOutputDrivers()' here? Is it 1 or not? If it is 1. then we are not testing that there is a file per driver.

gggrace14 replied:
Yes, task->numOutputDrivers() is now always 1 for Hive connector, after debugging into the code. So for tests I will check `EXPECT_EQ(countRecursiveFiles(*iterPartitionDirectory), 1)` for now.

(Originally I thought it's easier to switch from 1 to 1+ max drivers. It turns out it needs more code changes.)

mbasmanova commented on velox/exec/tests/TableWriteTest.cpp:
```diff
@@ -228,40 +291,258 @@ TEST_F(TableWriteTest, constantVectors) {
       makeHiveConnectorSplits(outputDirectory),
       "SELECT * FROM tmp");
+
+  // Verify that files written to each partition directory is per driver.
```
Same question as above.

mbasmanova commented on velox/exec/tests/TableWriteTest.cpp:
```diff
@@ -228,40 +291,258 @@ TEST_F(TableWriteTest, constantVectors) {
+
+  CursorParameters params;
+  params.planNode = plan;
+  params.queryCtx = std::make_shared<QueryCtx>(
```
Use AssertQueryBuilder

gggrace14 replied:
This is connector level config (rather than query config) and AssertQueryBuilder right now doesn't support connectorConfig. I'm adding connectorConfig to AssertQueryBuilder easily. But it eventually needs to add a setConnectorConfigOverridesUnsafe() API to QueryCtx.h. Let me know if you think it's not worthwhile and let me revert.

mbasmanova replied:
> I'm adding connectorConfig to AssertQueryBuilder

Thanks. These changes look good.

mbasmanova reviewed (commented):
@gggrace14 Ge, thank you for iterating on this PR. It is shaping up nicely. Overall looks great % a few comments below.

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -103,11 +161,37 @@ std::vector<std::string> HiveDataSink::finish() const {
 void HiveDataSink::close() {
   for (const auto& writer : writers_) {
-    writer->close();
+    if (writer) {
```
You're right.

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -103,11 +161,37 @@ std::vector<std::string> HiveDataSink::finish() const {
 
-void HiveDataSink::createWriter(
+void HiveDataSink::appendWriter(
+    vector_size_t partitionId,
```
Yeah, looks better, making it truly append.

gggrace14 commented on velox/connectors/hive/HivePartitionUtil.h:
```diff
@@ -19,12 +19,18 @@
+/// Make the partition name in the Hive style, with the partition values at row.
+/// Partition values are from the partition channels. Partition name is in
+/// the format of key1=value1/key2=value2/...
+std::string makePartitionName(
```
Removed

gggrace14 commented on velox/connectors/hive/tests/PartitionIdGeneratorTest.cpp:
```diff
@@ -139,27 +139,6 @@ TEST_F(PartitionIdGeneratorTest, stableIdsMultipleKeys) {
   }
 }
 
-TEST_F(PartitionIdGeneratorTest, maxPartitionId) {
```
Added

gggrace14 commented on velox/connectors/hive/HiveConfig.h:
```diff
@@ -31,8 +31,14 @@ class HiveConfig {
   static constexpr const char* kInsertExistingPartitionsBehavior =
       "insert_existing_partitions_behavior";
 
+  /// Maximum number of partitions per writer.
```
Hi Masha, yes, by "per writer" it means. "per a single instance of TableWriter operator" (or thread). I will change the comment and configs.rst. I copied the text from Java, and this is how I understand it for now. 

We're replicating this config from Java. Regarding whether it's 100 partitions per Task or per thread, my understanding for now is per thread, so the theoretic total of 500 partitions is true assuming my understanding is correct. The following is the code flow, in case you have time to help double check.

This config is validated in class HivePageSink. (`pagePartitioner` is like our `PartitionIdGenerator` class that is applied across inputs)
https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java#L470-L473

`pagePartitioner` is per `HivePageSink` instance which is per `TableWriterOperator` instance. I don't find `pageSink` field of `TableWriterOperator` class is shared among drivers/threads.
https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/TableWriterOperator.java#L137-L147,
https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/TableWriterOperator.java#L341

`LocalExecutionPlanner` class is where the number of drivers for `TableWriterOperator` is set.
https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java#L2590-L2595

mbasmanova replied:
@gggrace14 Ge, thank you for checking. I'm seeing that number of TableWriter operator instances is controlled by a session property task_partitioned_writer_count which defaults to the value of another session property task_writer_count.

```
                new PropertyMetadata<>(
                        TASK_PARTITIONED_WRITER_COUNT,
                        "Number of writers per task for partitioned writes. If not set, the number set by task.writer-count will be used",
                        BIGINT,
                        Integer.class,
                        taskManagerConfig.getPartitionedWriterCount(),
                        false,
                        value -> validateValueIsPowerOfTwo(value, TASK_PARTITIONED_WRITER_COUNT),
                        value -> value),
```

, but there is also a comment that suggests that writes are serialized. We'll need to dig in to understand the behavior better.

```
        @Override
        public PhysicalOperation visitTableWriter(TableWriterNode node, LocalExecutionPlanContext context)
        {
            // Set table writer count
            if (node.getTablePartitioningScheme().isPresent()) {
                context.setDriverInstanceCount(getTaskPartitionedWriterCount(session));
            }
            else {
                context.setDriverInstanceCount(getTaskWriterCount(session));
            }

            // serialize writes by forcing data through a single writer
            PhysicalOperation source = node.getSource().accept(this, context);
```

gggrace14 commented on velox/exec/tests/TableWriteTest.cpp:
```diff
@@ -39,13 +44,68 @@ class TableWriteTest : public HiveConnectorTestBase {
+  }
+
+  std::vector<std::shared_ptr<connector::ConnectorSplit>>
+  makeHiveConnectorSplitsWithNoTemporary(
```
Removed as the test cases do not need it anymore.

gggrace14 commented on velox/exec/tests/TableWriteTest.cpp:
```diff
@@ -228,40 +291,258 @@ TEST_F(TableWriteTest, constantVectors) {
+                    .planNode();
+
+    assertQuery(plan, "SELECT count(*) FROM tmp");
+    assertQuery(
```
The test case does not need it anymore.

gggrace14 commented on velox/exec/tests/TableWriteTest.cpp:
```diff
@@ -228,40 +291,258 @@ TEST_F(TableWriteTest, constantVectors) {
+    assertQuery(plan, "SELECT count(*) FROM tmp");
+    assertQuery(
+        PlanBuilder().tableScan(rowType_).planNode(),
+        makeHiveConnectorSplitsWithNoTemporary(outputDirectory),
```
Revised as suggested.

gggrace14 commented on velox/exec/tests/TableWriteTest.cpp:
```diff
@@ -39,13 +44,68 @@ class TableWriteTest : public HiveConnectorTestBase {
+  }
+
+  std::string partitionNameToPredicate(const std::string& partitionName) {
+    return std::regex_replace(partitionName, std::regex("/"), " AND ");
```
The func used to work only for non-strings, as it's customized for this specific test case. I'm making the func a bit general and it needs partition type info to tell between string and non-string.

gggrace14 commented:
Hi @mbasmanova , I have addressed the recent comments and uploaded a new revision. Ready for review. Would you be able to help take a look again?

mbasmanova reviewed (approved):
@gggrace14 Looks good to me % small comments.

gggrace14 commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -107,7 +165,28 @@ void HiveDataSink::close() {
+    writers_.reserve(numPartitions);
+    writerInfo_.reserve(numPartitions);
+    for (auto id = numWriters; id < numPartitions; id++) {
+      appendWriter(partitionIdGenerator_->partitionName(id));
```
Yes, the partition keys in partition name appear in the order of partition columns in the table schema. This is also how Presto Java workers assemble the partition name, so I think it's most safe to do the same. I document this in PartitionIdGenerator, as that is where the above order is enforce.

(I guess no particular order should be assumed by anywhere, while we might want to keep Velox and Java consistent to be safe. The layout of file system directory should be the source of truth for read.)

mbasmanova replied:
@gggrace14 I think we need to document this in HiveInsertTableHandle class. Velox doesn't have access to metastore and therefore cannot independently find out what is the table schema. It relies on the order of columns in HiveInsertTableHandle and assumes it matches table schema. Let's document this assumption clearly.
